File: crates/audio/src/lookup/musicbrainz.rs
use anyhow::Result;
use serde::Deserialize;

#[derive(Debug, Deserialize)]
struct SearchResponse {
    recordings: Vec<Recording>,
}

#[derive(Debug, Deserialize)]
struct Recording {
    id: String,
    title: String,
    #[serde(default)]
    length: Option<u64>,
    #[serde(default)]
    score: u32,
    #[serde(default, rename = "artist-credit")]
    artist_credit: Vec<ArtistCredit>,
    #[serde(default, rename = "release-list")]
    releases: Vec<Release>,
    #[serde(default)]
    disambiguation: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ArtistCredit {
    name: String,
}

#[derive(Debug, Deserialize)]
struct Release {
    title: String,
}

#[derive(Debug)]
pub struct MusicBrainzResult {
    pub title: String,
    pub artist: Option<String>,
    pub album: Option<String>,
    pub mbid: String,
    pub duration_ms: Option<u64>,
    pub duration: Option<String>,
}

pub async fn search_recording(title: &str, artist: Option<&str>) -> Result<Option<MusicBrainzResult>> {
    let query = match artist {
        Some(a) => format!("\"{}\" AND artist:\"{}\"", title, a),
        None => title.to_string(),
    };

    let url = format!(
        "https://musicbrainz.org/ws/2/recording/?query={}&limit=10&fmt=json",
        super::deezer::urlencod(&query),
    );

    let client = reqwest::Client::new();
    let resp: SearchResponse = client
        .get(&url)
        .header("User-Agent", "clickbAIt/0.1.0 (https://github.com/clickbait)")
        .send()
        .await?
        .json()
        .await?;

    let recordings = &resp.recordings;
    if recordings.is_empty() {
        return Ok(None);
    }

    // Prefer non-live recordings
    let rec = recordings
        .iter()
        .find(|r| {
            r.disambiguation
                .as_deref()
                .map_or(true, |d| !d.to_lowercase().contains("live"))
        })
        .unwrap_or(&recordings[0]);

    let artist_name = rec.artist_credit.first().map(|a| a.name.clone());
    let album = rec.releases.first().map(|r| r.title.clone());

    let duration_str = rec.length.map(|ms| {
        let total_secs = ms / 1000;
        format!("{}:{:02}", total_secs / 60, total_secs % 60)
    });

    Ok(Some(MusicBrainzResult {
        title: rec.title.clone(),
        artist: artist_name,
        album,
        mbid: rec.id.clone(),
        duration_ms: rec.length,
        duration: duration_str,
    }))
}

File: src/authoring/lookup/musicbrainz.ts
/** MusicBrainz API adapter.
 *
 *  Prefers non-"live" recordings via the `disambiguation` field since the
 *  original Rust impl found Deezer BPMs weren't useful for live cuts.
 *
 *  Requires a descriptive User-Agent per MB policy.
 */

import type { MusicBrainzResult } from "./types.js";

interface Recording {
  id: string;
  title: string;
  length?: number;
  "artist-credit"?: Array<{ name: string }>;
  "release-list"?: Array<{ title: string }>;
  disambiguation?: string;
}

interface SearchResponse {
  recordings?: Recording[];
}

export async function searchRecording(
  title: string,
  artist?: string,
): Promise<MusicBrainzResult | null> {
  const query = artist
    ? `"${title}" AND artist:"${artist}"`
    : title;
  const url =
    `https://musicbrainz.org/ws/2/recording/?query=${encodeURIComponent(query)}` +
    `&limit=10&fmt=json`;

  const resp = await fetch(url, {
    headers: { "User-Agent": "clickbAIt/0.1.0 (https://github.com/clickbait)" },
  });
  const json = (await resp.json()) as SearchResponse;
  const recordings = json.recordings ?? [];
  if (recordings.length === 0) return null;

  const rec =
    recordings.find(
      (r) => !(r.disambiguation ?? "").toLowerCase().includes("live"),
    ) ?? recordings[0];

  const artistName = rec["artist-credit"]?.[0]?.name;
  const album = rec["release-list"]?.[0]?.title;
  const duration = rec.length
    ? `${Math.floor(rec.length / 60000)}:${String(
        Math.floor((rec.length / 1000) % 60),
      ).padStart(2, "0")}`
    : undefined;

  return {
    title: rec.title,
    artist: artistName,
    album,
    mbid: rec.id,
    durationMs: rec.length,
    duration,
  };
}
